tests/objvariant/tbranch_change.nim

```diff
@@ -0,0 +1,150 @@
+discard """
+  description: "Tests to make sure branch changing works correctly"
+  targets: "c cpp"
+  matrix: "--gc:refc"
```

Comment:
Might as well cover arc and orc here too, the parts that don't work under them can be put into a separate test with a `knowIssue` in its testament spec. It will presently disable the test, but in the future they'll be run and checked to see if they've been fixed.

Reply:
That would be more or less all tests then. `arc`/`orc` doesn't emit a branch reset for types that don't have a destructor (fixable, but would be another breaking change).

I could add a dummy `string` field to all variants in order to force `arc`/`orc` to generate the reset logic.

Reply:
> Maybe I'm being especially naive, but doesn't the lack of reset logic being emitted seem like a bug?

The intention behind how it's implemented with `arc` seems to be to only prevent memory leaks and crashes due to corruption of objects with destructors. Since variant objects without destructors don't (directly) exhibit the aforementioned problems, branches are simply not reset on change for those.

If branch switching is to be officially allowed then I'd say it is a bug.

> I'd prefer we flag the type perhaps and work off that if we can?

I'm not sure I understand what you mean by flagging the type? Putting the failing test blocks behind a `when not defined(gcArc)`?

Reply:
By flagging the type I meant a tyFlag on the type data, but I'm sure that has issues.

Reply:
Getting `injectdestructors` to also emit a branch reset for types without destructors only requires removing a single `if` check